Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm

```diff
@@ -705,15 +717,27 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
     auto sourceContentWorldType = senderParameters.contentWorldType;
 
     enumerateFramesAndNamespaceObjects([&](auto& frame, auto& namespaceObject) {
-        // Skip all frames that don't match if a target frame identifier is specified.
-        if (frameIdentifier && !matchesFrame(frameIdentifier.value(), frame))
-            return;
-
         // Don't send the event to any listeners in the sender's page.
         auto webPageProxyIdentifier = frame.page()->webPageProxyIdentifier();
         if (senderParameters.pageProxyIdentifier == webPageProxyIdentifier)
             return;
 
+        if (targetParameters) {
+            // Skip all frames / documents that don't match the target parameters.
+            auto& frameIdentifier = targetParameters.value().frameIdentifier;
+            if (frameIdentifier && !matchesFrame(frameIdentifier.value(), frame))
+                return;
+
+            if (auto& documentIdentifier = targetParameters.value().documentIdentifier) {
+                auto frameDocumentIdentifier = toDocumentIdentifier(frame);
+                if (!frameDocumentIdentifier)
+                    return;
+
+                if (documentIdentifier != frameDocumentIdentifier)
+                    return;
+            }
+        }
```

Comment:
Can this be shared in any way with the code in lines 631-646?

Reply:
Yeah I can make a helper function for that.